addons/report/models/report.py

```diff
@@ -198,73 +487,62 @@ def get_pdf(self, docids, report_name, html=None, data=None):
-                body = lxml.html.tostring(node)
-                header = render_minimal(dict(subst=True, body=body, base_url=base_url))
-                headerhtml.append(header)
-
-            for node in root.xpath(match_klass.format('footer')):
-                body = lxml.html.tostring(node)
-                footer = render_minimal(dict(subst=True, body=body, base_url=base_url))
-                footerhtml.append(footer)
-
-            for node in root.xpath(match_klass.format('article')):
-                # Previously, we marked some reports to be saved in attachment via their ids, so we
-                # must set a relation between report ids and report's content. We use the QWeb
-                # branding in order to do so: searching after a node having a data-oe-model
-                # attribute with the value of the current report model and read its oe-id attribute
-                if docids and len(docids) == 1:
-                    reportid = docids[0]
-                else:
-                    oemodelnode = node.find(".//*[@data-oe-model='%s']" % report.model)
-                    if oemodelnode is not None:
-                        reportid = oemodelnode.get('data-oe-id')
-                        if reportid:
-                            reportid = int(reportid)
-                    else:
-                        reportid = False
-
-                # Extract the body
-                body = lxml.html.tostring(node)
-                reportcontent = render_minimal(dict(subst=False, body=body, base_url=base_url))
-
-                contenthtml.append(tuple([reportid, reportcontent]))
-
-        except lxml.etree.XMLSyntaxError:
-            contenthtml = []
-            contenthtml.append(html)
-            save_in_attachment = {}  # Don't save this potentially malformed document
-
-        # Get paperformat arguments set in the root html tag. They are prioritized over
-        # paperformat-record arguments.
-        specific_paperformat_args = {}
-        for attribute in root.items():
-            if attribute[0].startswith('data-report-'):
-                specific_paperformat_args[attribute[0]] = attribute[1]
+        headerhtml, contenthtml, footerhtml, sorted_docids, specific_paperformat_args = self.with_context(context)._prepare_html(html, docids, report.model)
 
         # Run wkhtmltopdf process
-        return self._run_wkhtmltopdf(
+        documents_paths = run_wkhtmltopdf(
             headerhtml, footerhtml, contenthtml, context.get('landscape'),
-            paperformat, specific_paperformat_args, save_in_attachment,
+            paperformat, specific_paperformat_args,
             context.get('set_viewport_size'),
         )
 
+        # Look for attachment_name both in the context and parameter
+        attachment_name = attachment_name or context.get('attachment_name')
+
+        # Save in attachment
+        if attachment_name or report.attachment:
```

Comment:
@smetl Can you take a look at https://github.com/odoo/odoo/pull/15268/files
With your implementation, it's  not possible to post process the document. 
1. You only call the create_attachment method if the report doesn't exist.
2. You pass a copy of the generated report to the `create_attachment` method. Therefore if you enrich the generated report as proposed in the docstring of `create_attachment` these changes will not be part of the file returned by the method.
3. The logic to check if a generated report must be saved and to retrieve the saved one is no more extendable. (the method `_check_attachment_use` is removed)

Your changes are a good step forward in the cleanup of the report module. I hope that my comments can be taken into account.

Regards,

lmi

cc @sle-odoo